## Online Retail Dataset Cleaning

#### 1. Load and Review

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("../data/raw/online_retail.csv")
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           InvoiceDate  UnitPrice  CustomerID         Country  
0  2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1  2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2  2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3  2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4  2010-12-01 08:26:00       3.39     17850.0  United Kingdom

In [3]:
df_clean = df.copy()

In [4]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


---

#### 2. Drop Missing Values

In [5]:
df_clean.isnull().sum().sort_values(ascending=False)

CustomerID     135080
Description      1454
StockCode           0
InvoiceNo           0
Quantity            0
InvoiceDate         0
UnitPrice           0
Country             0
dtype: int64

For customer-level analysis, 135080 rows without `CustomerID` are unusable.

In [6]:
df_clean = df_clean.dropna(subset=["CustomerID"])

---

#### 3. Drop Duplicates

In [7]:
df_clean.duplicated().sum()

np.int64(5225)

In [8]:
duplicates = df_clean[df_clean.duplicated(keep=False)]
duplicates.head(10)

InvoiceNo StockCode                        Description  Quantity  \
485    536409     22111       SCOTTIE DOG HOT WATER BOTTLE         1   
489    536409     22866      HAND WARMER SCOTTY DOG DESIGN         1   
494    536409     21866        UNION JACK FLAG LUGGAGE TAG         1   
517    536409     21866        UNION JACK FLAG LUGGAGE TAG         1   
521    536409     22900    SET 2 TEA TOWELS I LOVE LONDON          1   
527    536409     22866      HAND WARMER SCOTTY DOG DESIGN         1   
537    536409     22900    SET 2 TEA TOWELS I LOVE LONDON          1   
539    536409     22111       SCOTTIE DOG HOT WATER BOTTLE         1   
548    536412     22327  ROUND SNACK BOXES SET OF 4 SKULLS         1   
555    536412     22327  ROUND SNACK BOXES SET OF 4 SKULLS         1   

             InvoiceDate  UnitPrice  CustomerID         Country  
485  2010-12-01 11:45:00       4.95     17908.0  United Kingdom  
489  2010-12-01 11:45:00       2.10     17908.0  United Kingdom  
494  2010-12-01 11:45:00       1.25     17908.0  United Kingdom  
517  2010-12-01 11:45:00       1.25     17908.0  United Kingdom  
521  2010-12-01 11:45:00       2.95     17908.0  United Kingdom  
527  2010-12-01 11:45:00       2.10     17908.0  United Kingdom  
537  2010-12-01 11:45:00       2.95     17908.0  United Kingdom  
539  2010-12-01 11:45:00       4.95     17908.0  United Kingdom  
548  2010-12-01 11:49:00       2.95     17920.0  United Kingdom  
555  2010-12-01 11:49:00       2.95     17920.0  United Kingdom

In [9]:
df_clean = df_clean.drop_duplicates()

In [10]:
df_clean.duplicated().sum()

np.int64(0)

---

#### 4. Check Negative or Zero Quantities

In [11]:
df_clean[df_clean["Quantity"] <= 0].head()

InvoiceNo StockCode                       Description  Quantity  \
141   C536379         D                          Discount        -1   
154   C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235   C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236   C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237   C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   

             InvoiceDate  UnitPrice  CustomerID         Country  
141  2010-12-01 09:41:00      27.50     14527.0  United Kingdom  
154  2010-12-01 09:49:00       4.65     15311.0  United Kingdom  
235  2010-12-01 10:24:00       1.65     17548.0  United Kingdom  
236  2010-12-01 10:24:00       0.29     17548.0  United Kingdom  
237  2010-12-01 10:24:00       0.29     17548.0  United Kingdom

In [12]:
(df_clean["Quantity"] <= 0).sum()

np.int64(8872)

In [13]:
(df_clean["Quantity"] <= 0).mean() * 100

np.float64(2.209141343213713)

Returns handling

Rows with `Quantity <= 0` represent product returns or cancelled transactions. As these records distort sales and revenue metrics, they were excluded from sales-focused analysis, while acknowledging that return behavior could be analyzed separately.

In [14]:
df_clean = df_clean[df_clean["Quantity"] > 0]

---

#### 5. Zero or Negative Prices

In [15]:
df_clean[df_clean["UnitPrice"] <= 0].head()

InvoiceNo StockCode                   Description  Quantity  \
9302     537197     22841  ROUND CAKE TIN VINTAGE GREEN         1   
33576    539263     22580  ADVENT CALENDAR GINGHAM SACK         4   
40089    539722     22423      REGENCY CAKESTAND 3 TIER        10   
47068    540372     22090       PAPER BUNTING RETROSPOT        24   
47070    540372     22553        PLASTERS IN TIN SKULLS        24   

               InvoiceDate  UnitPrice  CustomerID         Country  
9302   2010-12-05 14:02:00        0.0     12647.0         Germany  
33576  2010-12-16 14:36:00        0.0     16560.0  United Kingdom  
40089  2010-12-21 13:45:00        0.0     14911.0            EIRE  
47068  2011-01-06 16:41:00        0.0     13081.0  United Kingdom  
47070  2011-01-06 16:41:00        0.0     13081.0  United Kingdom

In [16]:
(df_clean["UnitPrice"] <= 0).sum()

np.int64(40)

In this dataset, `UnitPrice == 0` usually means:

- Free items / samples
- Data entry issues
- Promotional adjustments

But for **revenue analysis**, they are **not meaningful sales**.

In [17]:
df_clean = df_clean[df_clean["UnitPrice"] > 0]

In [18]:
(df_clean["UnitPrice"] <= 0).sum()

np.int64(0)

---

#### 6. Preview Clean dataset

In [19]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 392692 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    392692 non-null  object 
 1   StockCode    392692 non-null  object 
 2   Description  392692 non-null  object 
 3   Quantity     392692 non-null  int64  
 4   InvoiceDate  392692 non-null  object 
 5   UnitPrice    392692 non-null  float64
 6   CustomerID   392692 non-null  float64
 7   Country      392692 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 27.0+ MB


In [20]:
df_clean.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           InvoiceDate  UnitPrice  CustomerID         Country  
0  2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1  2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2  2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3  2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4  2010-12-01 08:26:00       3.39     17850.0  United Kingdom

---

#### 7. Create Revenue Metric

In [21]:
df_clean["Revenue"] = df_clean["Quantity"] * df_clean["UnitPrice"]

---

#### 8. Convert Date

In [22]:
df_clean["InvoiceDate"] = pd.to_datetime(df_clean["InvoiceDate"])

---

#### 9. Sanity Check

In [23]:
df_clean.describe()

Quantity                    InvoiceDate      UnitPrice  \
count  392692.000000                         392692  392692.000000   
mean       13.119702  2011-07-10 19:13:07.771892480       3.125914   
min         1.000000            2010-12-01 08:26:00       0.001000   
25%         2.000000            2011-04-07 11:12:00       1.250000   
50%         6.000000            2011-07-31 12:02:00       1.950000   
75%        12.000000            2011-10-20 12:53:00       3.750000   
max     80995.000000            2011-12-09 12:50:00    8142.750000   
std       180.492832                            NaN      22.241836   

          CustomerID        Revenue  
count  392692.000000  392692.000000  
mean    15287.843865      22.631500  
min     12346.000000       0.001000  
25%     13955.000000       4.950000  
50%     15150.000000      12.450000  
75%     16791.000000      19.800000  
max     18287.000000  168469.600000  
std      1713.539549     311.099224

In [24]:
df_clean.sort_values("Quantity", ascending=False).head(10)

InvoiceNo StockCode                          Description  Quantity  \
540421    581483     23843          PAPER CRAFT , LITTLE BIRDIE     80995   
61619     541431     23166       MEDIUM CERAMIC TOP STORAGE JAR     74215   
421632    573008     84077    WORLD WAR 2 GLIDERS ASSTD DESIGNS      4800   
206121    554868     22197                 SMALL POPCORN HOLDER      4300   
97432     544612     22053                EMPIRE DESIGN ROSETTE      3906   
270885    560599     18007  ESSENTIAL BALM 3.5g TIN IN ENVELOPE      3186   
52711     540815     21108   FAIRY CAKE FLANNEL ASSORTED COLOUR      3114   
160546    550461     21108   FAIRY CAKE FLANNEL ASSORTED COLOUR      3114   
433788    573995     16014          SMALL CHINESE STYLE SCISSOR      3000   
291249    562439     84879        ASSORTED COLOUR BIRD ORNAMENT      2880   

               InvoiceDate  UnitPrice  CustomerID         Country    Revenue  
540421 2011-12-09 09:15:00       2.08     16446.0  United Kingdom  168469.60  
61619  2011-01-18 10:01:00       1.04     12346.0  United Kingdom   77183.60  
421632 2011-10-27 12:26:00       0.21     12901.0  United Kingdom    1008.00  
206121 2011-05-27 10:52:00       0.72     13135.0  United Kingdom    3096.00  
97432  2011-02-22 10:43:00       0.82     18087.0  United Kingdom    3202.92  
270885 2011-07-19 17:04:00       0.06     14609.0  United Kingdom     191.16  
52711  2011-01-11 12:55:00       2.10     15749.0  United Kingdom    6539.40  
160546 2011-04-18 13:20:00       2.10     15749.0  United Kingdom    6539.40  
433788 2011-11-02 11:24:00       0.32     16308.0  United Kingdom     960.00  
291249 2011-08-04 18:06:00       1.45     12931.0  United Kingdom    4176.00

In [25]:
df_clean[df_clean["Quantity"] > 10000][["Quantity", "UnitPrice", "Revenue"]].head()

Quantity  UnitPrice   Revenue
61619      74215       1.04   77183.6
540421     80995       2.08  168469.6

➤ High-quantity transactions

The dataset contains a small number of transactions with very large quantities, resulting in a heavily right-skewed distribution. Upon inspection, these records correspond to bulk purchases with low unit prices and realistic total revenue values. Therefore, they were retained, as they represent legitimate business activity.

---

#### 10. Save Clean Dataset

In [26]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 392692 entries, 0 to 541908
Data columns (total 9 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    392692 non-null  object        
 1   StockCode    392692 non-null  object        
 2   Description  392692 non-null  object        
 3   Quantity     392692 non-null  int64         
 4   InvoiceDate  392692 non-null  datetime64[ns]
 5   UnitPrice    392692 non-null  float64       
 6   CustomerID   392692 non-null  float64       
 7   Country      392692 non-null  object        
 8   Revenue      392692 non-null  float64       
dtypes: datetime64[ns](1), float64(3), int64(1), object(4)
memory usage: 30.0+ MB


In [ ]:
df_clean.to_csv("../data/processed/online_retail_clean.csv", index=False)